### PR TITLE
feat: Add a new configuration key to inject custom credentials class for authentication 

### PR DESCRIPTION
**Pull request by vermas2012**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable-hbase/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+   */
+  public abstract CompletableFuture<Map<String, List<String>>> getRequestMetadata(
+      URI uri, Executor executor) throws IOException;
+
+  /** Returns the HBase configuration used to create this object. */
+  public Configuration getConfiguration() {
+    return configuration;
+  }
+
+  protected Configuration configuration;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java:68]: why is this necessary?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java [hunk 2]
```diff
@@ -548,6 +550,12 @@ private void configureCredentialProvider(StubSettings.Builder<?, ?> stubSettings
       stubSettings.setCredentialsProvider(
           FixedCredentialsProvider.create(
               buildCredentialFromPrivateKey(serviceAccount, keyFileLocation)));
+    } else if (!Strings.isNullOrEmpty(configuration.get(BIGTABLE_CUSTOM_CREDENTIALS_CLASS_KEY))) {
+      String customAuthClassName = configuration.get(BIGTABLE_CUSTOM_CREDENTIALS_CLASS_KEY);
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:554]: configuration.getClass(key, BigtableOAuthCredentials.class) and then update the method signatures appropriately

Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:554]: Also why does this have the least precedence in the auth priority? I would think it belongs right after  BigtableExtendedConfiguration

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    if (!BigtableOAuthCredentials.class.isAssignableFrom(authClass)) {
+      throw new IllegalArgumentException(
+          "Custom credentials class ["
+              + customAuthClassName
+              + "] must be a child of "
+              + BigtableOAuthCredentials.class.getName()
+              + ".");
+    }
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java:45]: this should be replaced with hadoop's Configuration#getClass()

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException(
+          "Custom credentials class ["
+              + customAuthClassName
+              + "] must implement a constructor with single argument of type "
+              + Configuration.class.getName()
+              + ".",
+          e);
+    }
+
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java:59]: This and the newInstance should probably go into a static method of the BigtableOAuthCredentials interface/ abstract class

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    try {
+      return new BigtableCredentialsWrapper(
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java:61]: then you can inline BigtableCredentialsWrapper into the VeneerSettings and get rid of this class

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  public BigtableCredentialsWrapper(BigtableOAuthCredentials btCredentials) {
+    this.bigtableCredentials = btCredentials;
+  }
+
+  private final BigtableOAuthCredentials bigtableCredentials;
+
+  @Override
+  public String getAuthenticationType() {
+    return "OAuth2";
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:43]: add a comment explaining where you got this magic string

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      return bigtableCredentials.getRequestMetadata(uri, MoreExecutors.directExecutor()).get();
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to fetch credentials.", e);
+    }
+  }
+
+  public void getRequestMetadata(
+      final URI uri, Executor executor, final RequestMetadataCallback callback) {
+    executor.execute(() -> blockingGetToCallback(uri, callback));
+  }
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:58]: This is quite messy and totally my fault. I think you were right that we should just use the blocking variant and be done with it....this is very hard to reason about.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Get the current request metadata.
+   *
+   * <p>This should be called by the transport layer on each request, and the data should be
+   * populated in headers or other context.
+   *
+   * <p>The convention for handling binary data is for the key in the returned map to end with
+   * {@code "-bin"} and for the corresponding values to be base64 encoded.
+   *
+   * <p>This class should handle caching and refeshing of the metadata associated with the request.
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java:56]: Add a comment that the ideally the refresh would be async and preemptive

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  /**
+   * Get the current request metadata.
+   *
+   * <p>This should be called by the transport layer on each request, and the data should be
+   * populated in headers or other context.
+   *
+   * <p>The convention for handling binary data is for the key in the returned map to end with
+   * {@code "-bin"} and for the corresponding values to be base64 encoded.
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java:54]: I am just copying what was in Credentials class. As this is pass through.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * A public interface that should be implemented by the users who want to inject a custom {@link
+ * com.google.auth.Credentials} implementation for auth purposes. Clients can't directly override
+ * the {@link com.google.auth.Credentials} class as it is shaded by Cloud Bigtable client.
+ *
+ * <p>Hence, customers should implement this class, which will be used for authentication. The
+ * authentication should be based on OAuth and must work by just including request metadata with
+ * each request at transport layer.
+ */
+public abstract class BigtableOAuthCredentials {
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java:35]: I want to keep the hbase configuration in the base class AND enforce the constructor with configuration. I think its ok to keep it as a abstract class.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java (hunk 8 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java:56]: done.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java (hunk 1 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuthCredentials.java:68]: If we don't pass the configuration object, there is no way to configure the class. The child classes will have to hardcode everything. 

For example, you have some regional auth server, how do you configure the server URL? Becuse this will be initialized in the HBase region server and not user's services. The best way to configure is to add a string to hbase-site.xml

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java (hunk 2 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java:554]: I don't think this is going to be a popular way to inject credentials. Customer will only use this method, when no other method will work. hence, its the last.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java (hunk 3 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java:45]: done.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java (hunk 4 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java:59]: agreed, with use of getClass, this method is simple enought to be a named construcutor itself.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java (hunk 5 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/util/CustomCredentialsProvider.java:61]: not sure what you mean by in-lining BigtableCredentialsWrapper into VeneerSettings?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java (hunk 7 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:58]: ack, reverting to the sync version only.

**Comment by vermas2012** [comment]
@igorbernstein2 PTAL.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  /**
+   * Get the current request metadata.
+   *
+   * <p>This should be called by the transport layer on each request, and the data should be
+   * populated in headers or other context.
+   *
+   * <p>The convention for handling binary data is for the key in the returned map to end with
+   * {@code "-bin"} and for the corresponding values to be base64 encoded.
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java:55]: This is not relevant for us. Instead this should explicitly name the keys that need to be specified for Bigtable service

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  /**
+   * Creates a new instance of a child of @{@link BigtableOAuth2Credentials}.
+   *
+   * @param bigtableAuthClass the child class to be instantiated
+   * @param conf HBase configuration required to configure the @bigtableAuthClass
+   * @return a new instance of @bigtableAuthClass
+   */
+  public static Credentials newInstance(
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java:77]: can this be package private or at the very least it should be marked internal

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+          "Failed to create object of custom Credentials class ["
+              + bigtableAuthClass.getName()
+              + "].",
+          e);
+    }
+  }
+
+  protected Configuration configuration;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java:105]: Its weird that you have both a public getter and protected member variable for Configuration. I think it would be a lot simpler if you drop both and let endusers retain the ref from the constructor

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A wrapper that implements @{@link Credentials} interface and delegates calls to the
+ * underlying @{@link BigtableOAuth2Credentials} object. This is required to decouple the users of
+ * shaded bigtable from the @{@link Credentials} class.
+ */
+public final class BigtableCredentialsWrapper extends Credentials {
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:30]: Please mark this as internal api

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+  public BigtableCredentialsWrapper(BigtableOAuth2Credentials btCredentials) {
+    this.bigtableCredentials = btCredentials;
+  }
+
+  private final BigtableOAuth2Credentials bigtableCredentials;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:36]: member variables should come before the ctor

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+  public boolean hasRequestMetadataOnly() {
+    return true;
+  }
+
+  @Override
+  public void refresh() throws IOException {
+    // No-op. btCredentials should refresh internally as required.
+  }
+
+  public BigtableOAuth2Credentials getBigtableCredentials() {
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:68]: Why do you need a getter?

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java (hunk 11 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java:55]: What are those keys?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java (hunk 12 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java:77]: marked as internal. If I make it package private I i will need to move the wrapper, which is at its right place right now.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java (hunk 13 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java:105]: ok.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java (hunk 16 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableCredentialsWrapper.java:68]: I was using it for testing, but yeah, we don't really need it. Removed.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java (hunk 11 above)
Comment (vermas2012) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java]: I have removed the comments about binary data.

**Comment by vermas2012** [comment]
@igorbernstein2 PTAL

**Review by igorbernstein2 — APPROVED** [review]
